File: examples/numerical_propagation/generic_dynamics_builder.rs
//! ```cargo
//! [dependencies]
//! brahe = { path = "../../" }
//! nalgebra = "0.33"
//! ```
//! Constructing a NumericalPropagator using the builder API.
//!
//! The builder takes the three required fields -- epoch, state, and
//! dynamics_fn -- directly as arguments to `builder()`. Optional fields
//! such as params and initial_covariance default when omitted and are
//! set through chained setters.

use brahe as bh;
use bh::traits::{DStatePropagator, DStateProvider};
use nalgebra as na;
use std::f64::consts::PI;

fn main() {
    bh::initialize_eop().unwrap();

    let epoch = bh::Epoch::from_datetime(2024, 1, 1, 12, 0, 0.0, 0.0, bh::TimeSystem::UTC);

    let omega = 2.0 * PI; // 1 Hz oscillation frequency
    let initial_state = na::DVector::from_vec(vec![1.0, 0.0]); // [x0, v0]

    let dynamics_fn: bh::DStateDynamics = Box::new(
        move |_t: f64, state: &na::DVector<f64>, params: Option<&na::DVector<f64>>| {
            let x = state[0];
            let v = state[1];
            let omega_sq = params.map(|p| p[0]).unwrap_or(omega * omega);
            na::DVector::from_vec(vec![v, -omega_sq * x])
        },
    );

    let mut prop = bh::DNumericalPropagator::builder(epoch, initial_state, dynamics_fn)
        .propagation_config(bh::NumericalPropagationConfig::default())
        .params(na::DVector::from_vec(vec![omega * omega]))
        .build()
        .unwrap();

    let period = 2.0 * PI / omega; // Period = 2*pi/omega = 1 second
    prop.propagate_to(epoch + 5.0 * period);

    let final_state = prop.state(epoch + 5.0 * period).unwrap();
    println!("Position after 5 periods: {:+.6} m", final_state[0]);
    println!("Velocity after 5 periods: {:+.6} m/s", final_state[1]);

    assert!((final_state[0] - 1.0).abs() < 0.01);
    println!("Example validated successfully!");
}

File: examples/numerical_propagation/generic_dynamics_builder.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Constructing a NumericalPropagator using the builder API.

The builder takes the three required fields -- epoch, state, and
dynamics_fn -- directly as arguments to builder(). Optional fields such
as params and initial_covariance default when omitted and are set
through chained setters.
"""

import numpy as np
import brahe as bh

bh.initialize_eop()

epoch = bh.Epoch.from_datetime(2024, 1, 1, 12, 0, 0.0, 0.0, bh.TimeSystem.UTC)

omega = 2.0 * np.pi  # 1 Hz oscillation frequency
initial_state = np.array([1.0, 0.0])  # [x0, v0]


def sho_dynamics(t, state, params):
    """Simple harmonic oscillator dynamics."""
    x, v = state[0], state[1]
    omega_sq = params[0] if params is not None else omega**2
    return np.array([v, -omega_sq * x])


prop = (
    bh.NumericalPropagator.builder(epoch, initial_state, sho_dynamics)
    .propagation_config(bh.NumericalPropagationConfig.default())
    .params(np.array([omega**2]))
    .build()
)

period = 2 * np.pi / omega  # Period = 2*pi/omega = 1 second
prop.propagate_to(epoch + 5 * period)

final_state = prop.current_state()
print(f"Position after 5 periods: {final_state[0]:+.6f} m")
print(f"Velocity after 5 periods: {final_state[1]:+.6f} m/s")

assert abs(final_state[0] - 1.0) < 0.01
print("Example validated successfully!")
